runtime/compiler/control/rossa.cpp

```diff
@@ -1868,13 +1868,24 @@ aboutToBootstrap(J9JavaVM * javaVM, J9JITConfig * jitConfig)
 #if defined(J9VM_OPT_SHARED_CLASSES)
    if (isSharedAOT)
       {
-      /* If AOT Shared Classes is turned ON, perform compatibility checks for AOT Shared Classes */
-      if (0) //!validateSharedClassAOTHeader(javaVM, curThread, compInfo))
+      /* If AOT Shared Classes is turned ON, perform compatibility checks for AOT Shared Classes
+       *
+       * This check has to be done after latePostProcessJIT so that all the necessary JIT options
+       * can be set
+       */
+      TR_J9SharedCacheVM *fe = (TR_J9SharedCacheVM *) TR_J9VMBase::get(jitConfig, curThread, TR_J9VMBase::AOT_VM);
+      if (!compInfo->reloRuntime()->validateAOTHeader(fe, curThread))
          {
-         TR::Options::getAOTCmdLineOptions()->setOption(TR_NoLoadAOT);
-         TR::Options::getAOTCmdLineOptions()->setOption(TR_NoStoreAOT);
-         TR::Options::setSharedClassCache(false);
-         TR_J9SharedCache::setSharedCacheDisabledReason(TR_J9SharedCache::AOT_DISABLED);
+         /* This could be the first run, in which case validateAOTHeader might return false because
```

Comment:
I wonder is there is any problem when there is no header, but the SCC has been created by another JVM, maybe because that JVM didn't have the time to generate any AOT compilations. Could IProfiler be present in such an SCC and therefore defeat the check here?

Reply:
Hm, that is definitely a possible scenario. Maybe I'd have to add code in the IProfiler to generate the header if it doesn't already exist in that case.

Reply:
At the moment, we only ever persist the IProfiler data either in `genIL` (if we haven't delayed AOT loads) or when we do an AOT load (if we have delayed AOT loads).

Do you think we should still have code to create a header in the IProfiler just in case?

Reply:
@mpirvu ^

Reply:
@dsouzai Yes, we still have a problem, because an AOT compilation starts, `genIL()` is performed and some IP data is stored in SCC, then, while the JIT compiler is in optimizer or codegen, the JVM enters the shutdown stage, the compilation is aborted and we don't reach the point where the header is written.

I do not see a problem if delayed AOT load feature is in place though (we must have compiled and stored an AOT body in order to reach the point where IP data is stored).

Reply:
Addressed in 0233e1d2ebdd5ba3beae46cb7d6b7f0a1c2f8b06